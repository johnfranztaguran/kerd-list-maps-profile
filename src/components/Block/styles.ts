import { StyleSheet, ViewStyle } from 'react-native';
import { themes, colors } from 'constant';

type Styles = {
  block: ViewStyle;
  row: ViewStyle;
  column: ViewStyle;
  card: ViewStyle;
  center: ViewStyle;
  middle: ViewStyle;
  left: ViewStyle;
  right: ViewStyle;
  top: ViewStyle;
  bottom: ViewStyle;
  shadow: ViewStyle;
  primary: ViewStyle;
  secondary: ViewStyle;
  gray1: ViewStyle;
  gray2: ViewStyle;
  black: ViewStyle;
  white: ViewStyle;
};

const styles: Styles = StyleSheet.create<Styles>({
  block: {
    flex: 1,
  },
  row: {
    flexDirection: 'row',
  },
  column: {
    flexDirection: 'column',
  },
  card: {
    borderRadius: 6,
  },
  center: {
    alignItems: 'center',
  },
  middle: {
    justifyContent: 'center',
  },
  left: {
    justifyContent: 'flex-start',
  },
  right: {
    justifyContent: 'flex-end',
  },
  top: {
    justifyContent: 'flex-start',
  },
  bottom: {
    justifyContent: 'flex-end',
  },
  shadow: {
    shadowColor: colors.black,
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 1,
    shadowRadius: 13,
    elevation: 2,
  },
  primary: { backgroundColor: themes.colors.primary },
  secondary: { backgroundColor: themes.colors.secondary },
  gray1: { backgroundColor: colors.gray1 },
  gray2: { backgroundColor: colors.gray2 },
  black: { backgroundColor: colors.black },
  white: { backgroundColor: colors.white },
});

export default styles;
