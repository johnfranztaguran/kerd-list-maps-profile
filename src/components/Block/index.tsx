import React from 'react';
import { View, Animated, ViewStyle, ViewProps } from 'react-native';
import { isNil } from 'ramda';
import styles from './styles';

type Props = {
  flex?: number | boolean;
  row?: boolean;
  column?: boolean;
  center?: boolean;
  middle?: boolean;
  left?: boolean;
  right?: boolean;
  top?: boolean;
  bottom?: boolean;
  card?: boolean;
  shadow?: boolean;
  color?: string;
  primary?: boolean;
  secondary?: boolean;
  gray1?: boolean;
  gray2?: boolean;
  black?: boolean;
  white?: boolean;
  space?:
    | 'space-between'
    | 'space-around'
    | 'space-evenly'
    | 'flex-end'
    | 'flex-start';
  padding?: number | Array<number>;
  margin?: number | Array<number>;
  animated?: boolean;
  wrap?: 'nowrap' | 'wrap' | 'wrap-reverse';
  children?: React.ReactNode;
  style?: ViewStyle | any[];
};

const Block = (props: Props & ViewProps): JSX.Element => {
  const {
    flex,
    row,
    column,
    center,
    middle,
    left,
    right,
    top,
    bottom,
    card,
    shadow,
    color,
    primary,
    secondary,
    gray1,
    gray2,
    black,
    white,
    space,
    padding,
    margin,
    animated,
    wrap,
    style,
    children,
    ...otherProps
  } = props;

  const handleMargins = (): object => {
    if (typeof margin === 'number') {
      return {
        marginTop: margin,
        marginRight: margin,
        marginBottom: margin,
        marginLeft: margin,
      };
    }

    if (typeof margin === 'object') {
      const marginSize = Object.keys(margin).length;
      switch (marginSize) {
        case 1:
          return {
            marginTop: margin[0],
            marginRight: margin[0],
            marginBottom: margin[0],
            marginLeft: margin[0],
          };
        case 2:
          return {
            marginTop: margin[0],
            marginRight: margin[1],
            marginBottom: margin[0],
            marginLeft: margin[1],
          };
        case 3:
          return {
            marginTop: margin[0],
            marginRight: margin[1],
            marginBottom: margin[2],
            marginLeft: margin[1],
          };
        default:
          return {
            marginTop: margin[0],
            marginRight: margin[1],
            marginBottom: margin[2],
            marginLeft: margin[3],
          };
      }
    }

    return {};
  };

  const handlePaddings = (): object => {
    if (typeof padding === 'number') {
      return {
        paddingTop: padding,
        paddingRight: padding,
        paddingBottom: padding,
        paddingLeft: padding,
      };
    }

    if (typeof padding === 'object') {
      const paddingSize = Object.keys(padding).length;
      switch (paddingSize) {
        case 1:
          return {
            paddingTop: padding[0],
            paddingRight: padding[0],
            paddingBottom: padding[0],
            paddingLeft: padding[0],
          };
        case 2:
          return {
            paddingTop: padding[0],
            paddingRight: padding[1],
            paddingBottom: padding[0],
            paddingLeft: padding[1],
          };
        case 3:
          return {
            paddingTop: padding[0],
            paddingRight: padding[1],
            paddingBottom: padding[2],
            paddingLeft: padding[1],
          };
        default:
          return {
            paddingTop: padding[0],
            paddingRight: padding[1],
            paddingBottom: padding[2],
            paddingLeft: padding[3],
          };
      }
    }

    return {};
  };

  const handleFlex = (): object => {
    if (flex) {
      return { flex };
    }
    return {};
  };

  const blockStyles = [
    styles.block,
    handleFlex(),
    flex === false && { flex: 0 }, // reset / disable flex
    row && styles.row,
    column && styles.column,
    center && styles.center,
    middle && styles.middle,
    left && styles.left,
    right && styles.right,
    top && styles.top,
    bottom && styles.bottom,
    !isNil(margin) && { ...handleMargins() },
    !isNil(padding) && { ...handlePaddings() },
    card && styles.card,
    shadow && styles.shadow,
    space && { justifyContent: space },
    wrap && { flexWrap: wrap },
    !isNil(color) && { backgroundColor: color }, // custom backgroundColor
    // predefined styles colors for backgroundColor
    primary && styles.primary,
    secondary && styles.secondary,
    gray1 && styles.gray1,
    gray2 && styles.gray2,
    black && styles.black,
    white && styles.white,
    style, // rewrite predefined styles
  ];

  if (animated && children) {
    return <Animated.View style={blockStyles}>{children}</Animated.View>;
  }

  return children ? (
    <View style={blockStyles} {...(otherProps as ViewProps)}>
      {children}
    </View>
  ) : (
    <View style={blockStyles} {...(otherProps as ViewProps)} />
  );
};

export default Block;
