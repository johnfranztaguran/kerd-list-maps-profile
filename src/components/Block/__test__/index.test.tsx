import React from 'react';
import { render } from '@testing-library/react-native';
import { Text as MockText } from 'react-native';
import Block from '../index';

const createTextProps = (props: any) => ({
  ...props,
});

describe('Component: Block', () => {
  it('Should render a <Block> with a chilren centered', () => {
    const props: any = createTextProps({ center: true, middle: true });
    const { baseElement } = render(
      <Block {...props}>
        <MockText>Test</MockText>
      </Block>,
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('Should render a <Block> with no children', () => {
    const props: any = createTextProps({});
    const { baseElement } = render(<Block {...props} />);
    expect(baseElement).toMatchSnapshot();
  });
});
