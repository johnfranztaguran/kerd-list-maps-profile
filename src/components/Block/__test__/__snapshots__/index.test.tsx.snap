// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Block Should render a <Block> with a chilren centered 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {},
          false,
          undefined,
          undefined,
          Object {
            "alignItems": "center",
          },
          Object {
            "justifyContent": "center",
          },
          undefined,
          undefined,
          undefined,
          undefined,
          false,
          false,
          undefined,
          undefined,
          undefined,
          undefined,
          false,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
    >
      <Text>
        Test
      </Text>
    </View>
  </View>
</View>
`;

exports[`Component: Block Should render a <Block> with no children 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {},
          false,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          false,
          false,
          undefined,
          undefined,
          undefined,
          undefined,
          false,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
    />
  </View>
</View>
`;
